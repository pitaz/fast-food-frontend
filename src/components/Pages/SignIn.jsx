import React, { Component } from "react";
import { connect } from "react-redux";
import { Link, Redirect } from "react-router-dom";
import { bindActionCreators } from "redux";
import PropTypes from "prop-types";
import { validateSignInInput } from "../../validations/validations";
import signIn, { removeErrorMsg } from "../../actions/auth/signIn";
import SignInForm from '../UI/form/SignInForm.jsx';

export class SignIn extends Component {
  state = {
    email: '',
    password: '',
    errors: {}
  };

  onChange = event => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value,
        errors: newErrors
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  };

  onSubmit = e => {
    const { action } = this.props;
    e.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      action.signinUser(this.state);
    }
  };

  isValid = () => {
    const { errors, isValid } = validateSignInInput(this.state);
    if (!isValid) {
      this.setState({ errors, password: ""});
    }

    return isValid;
  };

  render() {
    const { auth, error } = this.props;
  
    if (auth) {
      return <Redirect to="/menu" />;
    }

    return (
      <div className="container">
        <div className="wrapper card">
        <SignInForm 
        {...this.state}
        {...this.props}
        onChange={this.onChange}
        onSubmit={this.onSubmit}
        /> 
        </div>
      </div>
    );
  }
}

SignIn.propTypes = {
  auth: PropTypes.bool,
  error: PropTypes.instanceOf(Object),
  action: PropTypes.instanceOf(Object)
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error
});

const matchDispatchToProps = dispatch => ({
  action: bindActionCreators(
    {
      signinUser: signIn
    },
    dispatch
  )
});

export default connect(
  mapStateToProps,
  matchDispatchToProps
)(SignIn);
