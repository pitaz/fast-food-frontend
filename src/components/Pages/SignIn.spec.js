/* eslint-disable react/no-children-prop */
/* eslint-disable no-undef */
import React from 'react';
import expect from 'expect';

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SignIn } from './SignIn.jsx';


let wrapper;
let mockedState;
let props;

describe('The Signin Component Test Suite', () => {
  beforeEach(() => {
  mockedState = {
    auth: {
      isAuthenticated: true,
      user: {
        id: 1,
        role: 'admin',
        email: 'peter@mail.com',
        iat: 1549688481,
        exp: 1549774881
      },
      error: {}
    },
  };
  props = {
    action: {
      deleteError: jest.fn(),
      signinUser: jest.fn(),
    },
    error: {},
    auth: false,
    isValid: jest.fn(),
  };
  });


  it('should match the snapshot', () => {
    wrapper = shallow(<SignIn {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render <SignInForm /> Component', () => {
    wrapper = shallow(<SignIn {...props} />);
    expect(wrapper.length).toBe(1);
    expect(wrapper.find(SignIn)).toBeDefined();
  });


  it('should test for onchange', () => { 
    wrapper = shallow(<SignIn {...props} />);

    const event = {
      target: { name: 'email', value: 'test@test.com' },
      preventDefault: () => { },
    };

    wrapper.instance().onChange(event);
    expect(wrapper.state().email).toEqual('test@test.com');
    expect(wrapper.instance().state.email).toEqual(event.target.value);
  });

  it('should test for onchange', () => { 
    wrapper = shallow(<SignIn {...props} />);

    const event = {
      target: { name: 'email', value: 'test@test.com' },
      preventDefault: () => { },
    };

    wrapper.instance().onChange(event);
    expect(wrapper.state().email).toEqual('test@test.com');
    expect(wrapper.instance().state.email).toEqual(event.target.value);
  });

  it('should test for onsubmit', () => { 
    wrapper = shallow(<SignIn {...props} />);

    const event = {
      target: { name: 'email', value: 'testtest.com' },
      preventDefault: () => { },
    };

    wrapper.instance().onChange(event);
    expect(wrapper.state().error).toEqual(undefined);
    expect(wrapper.instance().state.email).toEqual(event.target.value);
  });

 it('should test the onSubmit function', () => {
    const event = { preventDefault: () => {} };
    wrapper = shallow(<SignIn {...props} />);
    const form = wrapper.find('form');
    form.simulate('submit', event);
    const loginUser = jest.fn(mockedState);
    const promise = new Promise((resolve) => {
      resolve(wrapper.instance().onSubmit);
    });
    promise.then(() => expect(loginUser).toHaveBeenCalledTimes(1));
  });
});
