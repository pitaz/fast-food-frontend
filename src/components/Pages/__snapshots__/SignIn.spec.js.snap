// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Signin Component Test Suite should match the snapshot 1`] = `
<div
  className="container"
>
  <div
    className="wrapper card"
  >
    <form>
      <div
        className="card-header"
      >
        <p>
          Sign In
        </p>
      </div>
      <div
        className="card-content form-group"
      >
        <label>
          <b>
            Email
          </b>
        </label>
        <input
          className="email"
          field="email"
          id="nnn"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          required={true}
          type="email"
          value=""
        />
        <p
          className="error"
        />
        <label>
          <b>
            Password
          </b>
        </label>
        <input
          field="password"
          name="password"
          onChange={[Function]}
          placeholder="Enter Password"
          required={true}
          type="password"
          value=""
        />
        <p
          className="error"
        />
        <button
          onClick={[Function]}
          type="submit"
        >
          Submit 
        </button>
        <h5>
          Already Registered? 
          <Link
            replace={false}
            to="/signup"
          >
            Sign Up
          </Link>
        </h5>
      </div>
    </form>
  </div>
</div>
`;
